### PR TITLE
Add an authorizing proxy for a backing Kube

### PR DESCRIPTION
**Pull request by smarterclayton**
Need to set our credentials onto the underlying request still.

### CODE REVIEW

**Comment by smarterclayton** [comment]
This needs an upstream change to split our registration into two restful.Containers, one for secured and one for not.

**Comment by smarterclayton** [comment]
[test] @liggitt @deads2k review

File: pkg/cmd/server/origin/master.go [hunk 1]
```diff
@@ -305,14 +296,30 @@ func (c *MasterConfig) InstallAPI(container *restful.Container) []string {
...
 		case OpenShiftAPIPrefixV1Beta1:
 			svc.Doc("OpenShift REST API, version v1beta1").ApiVersion("v1beta1")
+
+			// add the current user filter
+			// TODO: factor this better
+			filter := currentUserContextFilter(c.getRequestsToUsers())
+			routes := svc.Routes()
+			for i := range routes {
+				route := &routes[i]
+				if route.Method == "GET" && (route.Path == OpenShiftAPIPrefixV1Beta1+"/ns/{namespace}/users/{name}" || route.Path == OpenShiftAPIPrefixV1Beta1+"/users/{name}") {
```
Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: This looks a little weird.  Is this restricting me to **only** being able to do a get for my own user?  I would have expected that requesting a different user's information would be restricted by some special policy handler, not via url rewriting.

File: pkg/cmd/server/origin/master.go [hunk 2]
```diff
@@ -305,14 +296,30 @@ func (c *MasterConfig) InstallAPI(container *restful.Container) []string {
 	}
 
 	var root *restful.WebService
+	installed := 0
```
Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: There's a lot of installing happening in this method.  How about something a little more specific userFiltersApplied or maybe, 'theNumberOfTimesIAppliedTheCurrentUserContextFilterToARouteInMyServerForSecurityReasons'?

File: pkg/cmd/server/origin/master.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/cmd/server/origin/master.go]: This is adding a filter in front of getting any user that looks for ~ and transforms it.

**Comment by smarterclayton** [comment]
Did you have other comments?

File: pkg/cmd/server/origin/master.go (hunk 1 above)
Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: I'm not seeing it.  Is currentUserContextFilter supposed to be checking for a '~' before doing a replacement?

Comment (smarterclayton) [inline pkg/cmd/server/origin/master.go]: Good catch, copy and paste fail.

----- Original Message -----

Comment (smarterclayton) [inline pkg/cmd/server/origin/master.go]: Which highlights how abysmal our tests of auth and the master are right now.  Follow up item we need to come back to is that test-cmd needs to be validating auth, test-integration needs to be reusing the master code (not inventing its own), and this stuff needs to be unit tested more effectively.

----- Original Message -----

**Comment by deads2k** [comment]
lgtm

**Pull request merged**